rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/core/ReconnectableRSocket.kt

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.kotlin.core
+
+import io.ktor.utils.io.core.*
+import io.rsocket.kotlin.*
+import io.rsocket.kotlin.payload.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+
+internal typealias ReconnectPredicate = suspend (cause: Throwable, attempt: Long) -> Boolean
+
+@Suppress("FunctionName")
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
+internal fun ReconnectableRSocket(
+    connect: suspend () -> RSocket,
+    predicate: ReconnectPredicate,
+): RSocket {
+    val state = MutableStateFlow<ReconnectState>(ReconnectState.Connecting)
+
+    val job =
+        connect.asFlow()
+            .map<RSocket, ReconnectState> { ReconnectState.Connected(it) } //if connection established - state = connected
+            .onStart { emit(ReconnectState.Connecting) } //init - state = connecting
+            .retryWhen { cause, attempts -> predicate(cause, attempts) } //reconnection logic
+            .catch { emit(ReconnectState.Failed(it)) } //reconnection failed - state = failed
+            .mapNotNull {
+                println("STATE: $it")
+                state.value = it //set state //TODO replace with Flow.stateIn when coroutines 1.4.0 will be released
+
+                when (it) {
+                    is ReconnectState.Connected -> it.rSocket.join()  //await for connection completion
+                    is ReconnectState.Failed -> throw it.error //reconnect failed, cancel job
+                    ReconnectState.Connecting -> null //ignore, should never happen
+                }
+            }
+            .launchRestarting() //reconnect if old connection completed/failed
+
+    return ReconnectableRSocket(job, state)
+}
+
+private fun Flow<*>.launchRestarting(): Job = GlobalScope.launch(Dispatchers.Unconfined) {
+    while (isActive) {
+        try {
+            collect()
+        } catch (e: Throwable) {
+            // KLUDGE: K/N
+            cancel("Reconnect failed", e)
+            break
+        }
+    }
+}
+
+private sealed class ReconnectState {
+    object Connecting : ReconnectState()
+    data class Failed(val error: Throwable) : ReconnectState()
+    data class Connected(val rSocket: RSocket) : ReconnectState()
+}
+
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
+private class ReconnectableRSocket(
+    override val job: Job,
+    private val state: StateFlow<ReconnectState>,
+) : RSocket {
+
+    private val reconnectHandler = state.mapNotNull { it.handleState { null } }.take(1)
+
+    //null pointer will never happen
```

Comment:
Why not make the types non-null? Why allow this to be nullable in the type system?

Reply:
Because `handleState` uses upper in `flow` with null to skip values. I wanted to not repeat state handling logic.